### ISSUE TITLE
Out of Memory

### ISSUE THREAD

**Issue by mrudling** [docusign/docusign-node-client#183]
I have encountered an issue downloading PDFs and running out of memory.  The stack trace is below.

The api is called with:

```
envelopesApi::getDocument(<accountId>, <envelopeId>, 'combined', null, <callback>)
```

The memory issue occurs when the ApiClient attempts to deserialize the PDF which has a returnType of `Object`. Therefore, a potential fix for this issue would be to check if the return type is not `Object` because the error occurs when the if statement attempts to do Object.keys(data) which is the response.body of the PDF.

```
    if (returnType !== Object && (data == null || !Object.keys(data).length)) {
      // SuperAgent does not always produce a body; use the unparsed response as a fallback
      data = response.text;
    }
```

STACK TRACE

--- Last few GCs --->
[emulators] >
[emulators] >  [36403:0x102654000]    10254 ms: Mark-sweep 106.4 (133.6) -> 101.9 (119.1) MB, 10.6 / 0.0 ms  (+ 31.7 ms in 37 steps since start of marking, biggest step 2.5 ms, walltime since start of marking 49 ms) (average mu = 0.993, current mu = 0.995) finalize incr[36403:0x102654000]    11878 ms: Mark-sweep 265.0 (278.6) -> 262.3 (304.1) MB, 16.5 / 0.0 ms  (+ 5.4 ms in 2 steps since start of marking, biggest step 4.1 ms, walltime since start of marking 1162 ms) (average mu = 0.991, current mu = 0.986) allocation fa
[emulators] >
[emulators] >  <--- JS stacktrace --->
[emulators] >
[emulators] >  ==== JS stack trace =========================================
[emulators] >
[emulators] >      0: ExitFrame [pc: 0x13f23e5dbe3d]
[emulators] >  Security context: 0x14b7e619e6e9 <JSObject>
[emulators] >      1: keys [0x14b7e61851a1](this=0x14b7e6184639 <JSFunction Object (sfi = 0x14b7ef78e301)>,0x14b74ad07191 <Very long string[22711550]>)
[emulators] >      2: deserialize [0x14b717ffaef1] [/Users/<username>/v/<appname>/packages/<packagename>/functions/node_modules/docusign-esign/src/ApiClient.js:343] [bytecode=0x14b7a371e8c9 offset=66](this=0x14b7a705a381 <exports map = ...
[emulators] >
[emulators] >  FATAL ERROR: invalid table size Allocation failed - JavaScript heap out of memory
[emulators] >   1: 0x10003c597 node::Abort() [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   2: 0x10003c7a1 node::OnFatalError(char const*, char const*) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   3: 0x1001ad575 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   4: 0x100579242 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   5: 0x1006e4e81 v8::internal::OrderedHashTable<v8::internal::OrderedHashSet, 1>::Rehash(v8::internal::Handle<v8::internal::OrderedHashSet>, int) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   6: 0x1006e4175 v8::internal::OrderedHashSet::Add(v8::internal::Handle<v8::internal::OrderedHashSet>, v8::internal::Handle<v8::internal::Object>) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   7: 0x10065a31d v8::internal::KeyAccumulator::AddKey(v8::internal::Handle<v8::internal::Object>, v8::internal::AddKeyConversion) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   8: 0x10050e709 v8::internal::(anonymous namespace)::ElementsAccessorBase<v8::internal::(anonymous namespace)::FastStringWrapperElementsAccessor, v8::internal::(anonymous namespace)::ElementsKindTraits<(v8::internal::ElementsKind)9> >::CollectElementIndices(v8::internal::Handle<v8::internal::JSObject>, v8::internal::Handle<v8::internal::FixedArrayBase>, v8::internal::KeyAccumulator*) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >   9: 0x10065c2b7 v8::internal::KeyAccumulator::CollectOwnElementIndices(v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::Handle<v8::internal::JSObject>) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >  10: 0x10065b5ae v8::internal::KeyAccumulator::CollectOwnKeys(v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::Handle<v8::internal::JSObject>) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >  11: 0x10065aa4b v8::internal::KeyAccumulator::CollectKeys(v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::Handle<v8::internal::JSReceiver>) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]
[emulators] >  12: 0x10065a103 v8::internal::KeyAccumulator::GetKeys(v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::KeyCollectionMode, v8::internal::PropertyFilter, v8::internal::GetKeysConversion, bool, bool) [/Users/<username>/.nvm/versions/node/v10.15.3/bin/node]

**Comment by acooper4960** [comment]
Hi @mrudling DocuSign has a size limit for documents, what is the size of your PDF? https://developers.docusign.com/esign-rest-api/guides/resource-limits

**Comment by mrudling** [comment]
Hi,

It is 22mb.  It was created and sent to docusign using this library, then failed on fetching it.

**Comment by vasyldemchuk** [comment]
Hi, 

I have exactly the same problem, so I would extremely vote to merge given fix to master and release new version as soon as possible.

Thanks

**Comment by LarryKlugerDS** [comment]
I've duplicated the problem internally and have filed internal bug report DCM-4415

**Comment by acooper4960** [comment]
@mrudling this issue has been resolved in versions v5.8.1 and v4.10.1
